opencog/attentionbank/AttentionSCM.cc

```diff
@@ -0,0 +1,214 @@
+
+using namespace opencog;
+
+// ========================================================
+
+AttentionSCM::AttentionSCM(void)
+{
+	static bool is_init = false;
+	if (is_init) return;
+	is_init = true;
+	scm_with_guile(init_in_guile, this);
+}
+
+
+void* AttentionSCM::init_in_guile(void* self)
+{
+	scm_c_define_module("opencog attention-bank", init_in_module, self);
+	scm_c_use_module("opencog attention-bank");
+	return NULL;
+}
+
+void AttentionSCM::init_in_module(void* data)
+{
+	AttentionSCM* self = (AttentionSCM*) data;
+	self->init();
+}
+
+/// This is called while (opencog attention-bank) is the current module.
+/// Thus, all the definitions below happen in that module.
+void AttentionSCM::init(void)
+{
+	define_scheme_primitive("cog-av", &AttentionSCM::get_av, this, "attention-bank");
+	define_scheme_primitive("cog-set-av!", &AttentionSCM::set_av, this, "attention-bank");
+	define_scheme_primitive("cog-inc-vlti!", &AttentionSCM::inc_vlti, this, "attention-bank");
+	define_scheme_primitive("cog-dec-vlti!", &AttentionSCM::dec_vlti, this, "attention-bank");
+
+	// Attentional-focus-related stuff
+	define_scheme_primitive("cog-update-af", &AttentionSCM::update_af, this, "attention-bank");
+	define_scheme_primitive("cog-af-size", &AttentionSCM::af_size, this, "attention-bank");
+	define_scheme_primitive("cog-set-af-size!", &AttentionSCM::set_af_size, this, "attention-bank");
+	define_scheme_primitive("cog-stimulate", &AttentionSCM::stimulate, this, "attention-bank");
+}
+
+AttentionSCM::~AttentionSCM()
+{
+}
+
+AttentionValuePtr AttentionSCM::get_av(const Handle& h)
+{
+	return opencog::get_av(h);
+}
+
+Handle AttentionSCM::set_av(const Handle& h, const AttentionValuePtr& av)
+{
+	AtomSpace* atomspace = SchemeSmob::ss_get_env_as("cog-set-av!");
+	attentionbank(atomspace).change_av(h, av);
+	return h;
+}
+
+Handle AttentionSCM::inc_vlti(const Handle& h)
+{
+	AtomSpace* atomspace = SchemeSmob::ss_get_env_as("cog-inc-vlti!");
+	attentionbank(atomspace).inc_vlti(h);
+	return h;
+}
+
+Handle AttentionSCM::dec_vlti(const Handle& h)
+{
+	AtomSpace* atomspace = SchemeSmob::ss_get_env_as("cog-dec-vlti!");
+	attentionbank(atomspace).dec_vlti(h);
+	return h;
+}
+
+/**
+ *   Return AttentionalFocus Size
+ **/
+int AttentionSCM::af_size(void)
+{
+	AtomSpace* atomspace = SchemeSmob::ss_get_env_as("cog-af-size");
+	return attentionbank(atomspace).get_af_size();
+}
+
+/**
+ * Set AttentionalFocus Size
+ */
+int AttentionSCM::set_af_size (int ssize)
+{
+	AtomSpace* atomspace = SchemeSmob::ss_get_env_as("cog-set-af-size!");
+
+	attentionbank(atomspace).set_af_size(ssize);
+	return attentionbank(atomspace).get_af_size();
+}
+
+/**
+ * Return the list of top n atoms in the AttentionalFocus or
+ * return all atoms in the AF if n is unspecified or is larger
+ * than the AF size.
+ */
+Handle AttentionSCM::update_af(int n)
+{
+	AtomSpace* atomspace = SchemeSmob::ss_get_env_as("cog-af");
+
+	Handle af_anchor = atomspace->add_node(ANCHOR_NODE,
```

Comment:
@linas Why would you want to record the list of atoms in the attention-focus, explicitly in the atomspace?